### DISCUSSION TITLE
CubicSpline / cubic_interp and follow-up work

### DISCUSSION THREAD

**Discussion by isorin** [BJReplay/ha-solcast-solar#96] [General]
I've tested the cubic_interp function that @autoSteve added to replace PchipInterpolator from scipy.interpolate.

Attached Python test code: [spline_test.zip](https://github.com/user-attachments/files/16325056/spline_test.zip)

## Data input for the functions
x =  [0.0, 1800.0, 3600.0, 5400.0, 7200.0, 9000.0, 10800.0, 12600.0, 14400.0, 16200.0, 18000.0, 19800.0, 21600.0, 23400.0, 25200.0, 27000.0, 28800.0, 30600.0, 32400.0, 34200.0, 36000.0, 37800.0, 39600.0, 41400.0, 43200.0, 45000.0, 46800.00.0, 55800.0, 57600.0, 59400.0, 61200.0, 63000.0, 64800.0, 66600.0, 68400.0, 70200.0, 72000.0, 73800.0, 75600.0, 77400.0, 79200.0, 81000.0, 82800.0, 84600.0, 86400.0]

y =  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0.025, 0.1069, 0.2597, 0.4366, 0.5477, 0.7426, 1.1552, 1.3159, 2.5428, 2.0347, 2.328, 2.936, 2.7023, 2.4922, 2.6485, 2.9187, 3.3253, 3.1058, 3.4755, 4.2293, 3.247, 3.2566, 3.3216, 3.2024, 2.7272, 2.158, 0.1504, 0.0803, 0.0123, 0, 0, 0, 0, 0]

## Performance evaluation (calculate interpolated curve for 1 day worth of data)
PchipInterpolator took 0.00058s
CubicSpline took 0.00031s
cubic_interp took 0.00059s

Performance is not bad considering scipy uses C/C++ code which should be really fast.
Note: I added a for loop and run each function 100 times, cubic_interp takes around 2x time as PchipInterpolator but still very good.

## Validate results (compare diff against CubicSpline from scipy)
[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, , 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]

Test code rounds to 5 decimal points and prints difference against CubicSpline from scipy - 0.0 for all data points, which is good.

## Follow-up work
I think we can start using cubic_interp() to calculate the interpolated curve for the current day and then use this data to calculate:
- remaining_forecast for today
- power_now, power_30min, power_60_mins
- forecast current_hour, next_hour

At the same time we should start calculating energy using "Trapezoidal rule":
![image](https://github.com/user-attachments/assets/d6c3b0eb-ad54-45f9-ab1c-3f3ab99ad422)
as opposed to "Midpoint rule" (which we currently use):
![image](https://github.com/user-attachments/assets/079e70a4-8557-451a-865d-2e06545d0f2d)
Reference: https://en.wikipedia.org/wiki/Riemann_sum#Trapezoidal_rule

We can apply the trapezoidal rule to the interpolated curve to make it more accurate (hopefuly).

What do people think?

**Comment by autoSteve** [comment]
In my mind I imagine the first (trapezoidal) chart more represents what the current interpolation is doing.

The forecast is saying that for a given half hour period, here's how much power we reckon you'll be making over that period, so these are arranged as stacked values along the timeline, making the assumption that the power will be constant for the period so can be considered 1/2 x Wh.

Much less data is used, which makes a further assumption. Just prior/prior period, prior period, this perod, next period.

```
x = [-1800, 0, 1800, 3600, ]
y = [_data[i-1][_data_field] + _data[i][_data_field], _data[i][_data_field], 0, -1 * _data[i+1][_data_field], ]
```

**Comment by isorin** [comment]
@autoSteve I think you are right about the forecast meaning.
I forgot that what we get from Solcast is the average power for each 30 minutes interval.
So the rule we use is actually correct! 😄 No need for trapezoidal rule.

I will investigate some more in relation to the interpolation but what we have for now is fine.

How is/was Tokyo?

**Reply by autoSteve** [reply]
> I forgot that what we get from Solcast is the average power for each 30 minutes interval.
> So the rule we use is actually correct! 😄 No need for trapezoidal rule.

I keep having my mind bent at times coming back to this thought. The Solcast prediction should represent the mid-ish-point of each half hour interval, being the average. I think. The code assumes the start of each period for interpolations (both old code and splines), and that is where my head breaks, thinking about how everything probably should be shifted by fifteen minutes.

But it's a forecast. It _will_ be wrong...

**Comment by autoSteve** [comment]
Tokyo is being proper mad.

Go-carts through the streets tomoz.

Mad.

I'm considering not coming home. Until it gets cold here... then I'm deffo home.

**Comment by autoSteve** [comment]
Damn. I'm home.

Damn, that was fun.

A second opinion/test would be valued for a change to interpolations, @isorin.

The branch "Post-v4.0.40" is an attempt at stamping out some weirdness with interpolations. Some have noted that the "remaining today" and the "expected total for today" can be mis-aligned slightly at the beginning of some days.

It's an annoyance really, as the difference is ~0.2kWh, with the forecasted remaining value being inflated. This should not happen, and it makes my eye twitch.

There were also some cases where negative values were presented, and I kludged that by limiting to > 0. That kludge is currently removed in this branch.

In an attempt to address these oddities, and also cater for potential future use cases I have changed the cubic spline to be fed with an entire day of data points. By doing so, the pchip() function can get any interpolated forecasted value for any given second past midnight.

What I think is happening is that the way I implemented the limited data points is likely throwing off the curve, especially at the start of the day, so having a whole day to chew on will probably avoid that.

I have made getting an interpolated value for everything the default. A summation of on-interval values will not be interpolated, but for all else where an off-interval is requested the pchip() function is used. I do believe this is where your head was heading.

**Comment by isorin** [comment]
Hi Steve,

Welcome back!

I looked at the code and here are my comments:

Yes, this is what I had in mind, to pass the entire data for the current day into cubic_interp.

To be efficient, I think we should do this only once, after refreshing the data from Solcast, and store the results.
We can then these cached results and avoid recalculating for each sensor update.

If you look at my attached spline_test.py, this is how I do it:

x0 = [ j for j in range(0, 86400, 300) ]
y0 = cubic_interp(x0, x, y)

Then, "y0" will contain interpolated values for each 5 minutes intervals (to coincide with each interval the sensors get updated so we can use "y0" as a cached response)

One more thing: "cubic_interp" likes one more value as input in (x, y) so send in 48+1 values to avoid a negative value for the last value of "y0".

That was my plan.

Also, I was thinking to use the "trapezoidal rule" only when calculating incomplete intervals (< 30 minutes) using interpolated values for each 5 minutes interval and use the average original Solcast value for all complete (30 minutes) intervals.
(Not sure this is even needed, only midpoint interpolated values for each 5 minutes interval should be enough)

**Reply by autoSteve** [reply]
> x0 = [ j for j in range(0, 86400, 300) ]
> y0 = cubic_interp(x0, x, y)

Ack.

**Reply by isorin** [reply]
Btw.
One other thing to explore here is to find out if cubic_interp could be split into two parts:

1st part to initialise it like this:
CublicInterp = cubic_interp(x, y)
where it would calculate the coefficients and whatever it needs to prepare it
(x, y) - is the Solcast data for the current day

2nd part would be to use it as before like this:
y0 = CubicInterp(x0)

This means that we could call the 1st part when data is updated (or the date changes)
and then the 2nd part as we do right now for each sensor.

**Reply by isorin** [reply]
This is how PchipInterpolator works.
I will look into this as well.

**Comment by isorin** [comment]
Btw. I captured the results in Excel from the old "PchipInterpolator" to see how it looks in a visualisation in Excel
(I think the cubic_interp produces fairly similar results)

![image](https://github.com/user-attachments/assets/2c35cc7d-5f38-47b8-bbeb-5547d1ac1727)

**Comment by autoSteve** [comment]
The logic for cubic spline over a period is completely messed up. The function for any given moment is fine.

Working on take 2 for the period calc...

**Comment by autoSteve** [comment]
New commit to the branch.

What this does is build two splines, and only when buildforecastdata() is called, so only on startup or forecast get. These contain calculated five-minute values.

There are two functions that populate two dictionaries with the various sites (+all) and estimateXX breakdowns for the current day. When regular calls are made the dict is addressed appropriately on five minute boundaries to pull the interpolated values without any calculation happening. If an interpolated value is not available, then the period value is used (example, if the custom hour sensor is set to eight hours, when it's presently 20:35 there will not be an interpolated end value available).

The two splines: One is "momentary" forecasts. The other is a reducing "remaining today", used to calc custom sensor/remaining today. The reducing forecast takes the sum of remaining intervals for the day for each "y" data point.

It's dark here, so all of this has not passed a real world test yet.

**Comment by isorin** [comment]
There, I managed to split cubic_interp into the two parts.
[cubic_spline.zip](https://github.com/user-attachments/files/16415695/cubic_spline.zip)

To use it:

Call
   cubicSplineInterp = CubicInterp(x, y)
and store "cubicSplineInterp" - then update it with new data

Then, when calculating sensors we can call:
"interpolate" for an array of input values (as before)
   y0 = cubicSplineInterp .interpolate(x0)
or "interpolate1" for a single value
   y0 = cubicSplineInterp .interpolate1(x0)

I created a new method called "interpolate1" which takes a single value (not an array) and it is optimized for such use.
Have a look at the code, I like the constructor code as it only stores the original values (48 values) and the new "interpolate1" method which should be faster. It depends of what we need.

Other notes:
- I looked over your latest code changes, it looks a bit complicated to do the 5 minutes interval pre/processing
- With the class that I split in two parts we can simplify it (I think)

**Reply by autoSteve** [reply]
I've not yet checked out whether that might simplify things.

There's a new commit to the branch, by the way, fixing much after real-world testing.

The building of the spline dictionaries is not that complex at the core. What does complicate the functions is the potential selection of sensor attributes, plus the need to build two spline sets for each attribute, one a reducing remaining total, the other for forecast moments. The use of the reducing spline is also further complicated because only the current day of 30-minute values is included, and things like the "custom x hours" sensor may have an end date in "tomorrow" or beyond (it has a max of 144 hours).

**Reply by isorin** [reply]
I don't think we need to worry about "Custom X hours" - if it's too big, the interpolation of the end interval matter less. We can do a best effort, interpolate if we have the data, if not, use the original (linear) approach.

The way I see this working is building "CubicInterp(x, y)" with a rolling window of time.
What we need it to cover is from "now" to "+X hours" - providing X is not too far in the future (I would say a couple of hours, not more)
We also need it to cover the two points for "now +30mins" and "now +60mins"
When building it, we need to give it 2 extra points of data on each side (so it can interpolate correctly) plus 2 more points in the future so we only need to rebuild it every 1h (or after the data is refreshed)
I don't think we even need a full day (current day) of data.

The easiest way to achieve this is to store a "timestamp" of last rebuild in the "CubicInterp" and the first sensor that tries to use it will check how old the timestamp is and rebuild it with a new window of time so we would always have interpolated data for the window of time we need.

I will try something, if I can find some time.

**Reply by autoSteve** [reply]
New commit.

An extra two data points are added to the splines to cover +30 +60. Good call.

I've left the spline as "whole day" (plus an hour...) because it works as is, and it's only calculated a few times per day, so no biggie for processing. Plus, doing remaining from now to end-of-day is going to get tricky with the reducing forecast for remaining for the day, and will do my head in because need to add up the prior periods and add on. Better to just build the whole reducing slide for the day at each forecast get. Simple works.

For the custom X sensor, ack. I've reverted longer periods to use a linear interp. for beyond spline, and this tests well so far and is way better than the current step ladder.

I've also suppressed the tendancy of a spline to "bounce" at the beginning and end of useful values. We were seeing "slightly larger remaining than total expected for day" in early morning, and also a "slightly less than zero/just above zero" bounce at the end of a day. This will be gone. (It was a matter of a handful of Watts, so a rounding error, yet made my eye twitch, and also twitched the eyes of others that noticed it...)

**Reply by autoSteve** [reply]
> we need to give it 2 extra points of data on each side (so it can interpolate correctly) plus 2 more points in the future

I'm not sure I agree. The extra points would just be more zeros for almost everybody, because middle of the night at both ends. Unless one is in the arctic circle in summer, and I'm not sure we have many, if any users in the arctic circle. Could be wrong...

Plus two more in the future beyond the two more already included at the end of the day? For why? More zeroes.

**Reply by isorin** [reply]
This is only if implementing a rolling window of time, not for the case we have right now which will be before/after midnight. (zeros).

**Comment by autoSteve** [comment]
PR #103 for the branch Post-v4.0.40 is approved and I think ready to go. Solves config directory location for some HA installs, adds the day splines, and corrects a bug in v4.0.40 where forecast now/+30/+60 were using the incorrect day data.

**Comment by autoSteve** [comment]
The python in you will probably be well pleased with this refactor, @isorin if not a 'rolling' spline... I switched off 'it works' mode tonight to do this code properly.

```python
    def get_spline(self, spline, st, xx, _data, df, reducing=False):
        for _data_field in df:
            if st > 0:
                y = [_data[st+i][_data_field] for i in range(0, len(self._spline_period))]
                if reducing:
                    # Build a decreasing set of forecasted values instead
                    y = [0.5 * sum(y[i:]) for i in range(0, len(self._spline_period))]
                spline[_data_field] = cubic_interp(xx, self._spline_period, y)
                self.sanitise_spline(spline, _data_field, xx, y, reducing=reducing)
            else: # The list slice was not found, so zero all values in the spline
                spline[_data_field] = [0] * (len(self._spline_period) * 6)

    def sanitise_spline(self, spline, _data_field, xx, y, reducing=False):
        for j in xx:
            i = int(j/300)
            # Suppress negative values
            if math.copysign(1.0, spline[_data_field][i]) < 0:
                spline[_data_field][i] = 0.0
            # Suppress spline bounce
            if reducing:
                if i+1 <= len(xx)-1 and spline[_data_field][i+1] > spline[_data_field][i]: spline[_data_field][i+1] = spline[_data_field][i]
            else:
                k = int(math.floor(j/1800))
                if k+1 <= len(y)-1 and y[k] == 0 and y[k+1] == 0: spline[_data_field][i] = 0.0
        # Shift right by fifteen minutes because 30-minute averages, padding as appropriate
        if reducing:
            spline[_data_field] = ([spline[_data_field][0]]*3) + spline[_data_field]
        else:
            spline[_data_field] = ([0]*3) + spline[_data_field]

    def build_splines(self, variant, reducing=False):
        """Cubic splines for interpolated inter-interval momentary or reducing estimates"""
        df = ['pv_estimate'] + (['pv_estimate10'] if self.options.attr_brk_estimate10 else []) + (['pv_estimate90'] if self.options.attr_brk_estimate90 else [])
        xx = [ i for i in range(0, 1800*len(self._spline_period), 300) ]
        st, _ = self.get_forecast_list_slice(self._data_forecasts, self.get_day_start_utc()) # Get start of day index

        variant['all'] = {}
        self.get_spline(variant['all'], st, xx, self._data_forecasts, df, reducing=reducing)
        if self.options.attr_brk_site:
            for site in self._sites:
                variant[site['resource_id']] = {}
                self.get_spline(variant[site['resource_id']], st, xx, self._data_forecasts, df, reducing=reducing)

    async def spline_moments(self):
        try:
            self.build_splines(self.fc_moment)
        except Exception as e:
            _LOGGER.debug('Exception in spline_moments(): %s', e)

    def get_moment(self, site, _data_field, t):
        try:
            return self.fc_moment['all' if site is None else site][self._data_field if _data_field is None else _data_field][int(t / 300)]
        except Exception as e:
            _LOGGER.debug('Exception in get_moment(): %s', e)
            return 0

    async def spline_remaining(self):
        try:
            self.build_splines(self.fc_remaining, reducing=True)
        except Exception as e:
            _LOGGER.debug('Exception in spline_remaining(): %s', e)

    def get_remaining(self, site, _data_field, t):
        try:
            return self.fc_remaining['all' if site is None else site][self._data_field if _data_field is None else _data_field][int(t / 300)]
        except Exception as e:
            _LOGGER.debug('Exception in get_remaining(): %s', e)
            return 0
```
